pkg/querier/queryrange/limits.go

```diff
@@ -508,3 +511,52 @@ func MinWeightedParallelism(ctx context.Context, tenantIDs []string, configs []c
 		)
 	})
 }
+
+// validates log entries limits
+func validateMaxEntriesLimits(req *http.Request, reqLimit uint32, limits Limits) error {
+	tenantIDs, err := tenant.TenantIDs(req.Context())
+	if err != nil {
+		return httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	maxEntriesCapture := func(id string) int { return limits.MaxEntriesLimitPerQuery(req.Context(), id) }
+	maxEntriesLimit := validation.SmallestPositiveNonZeroIntPerTenant(tenantIDs, maxEntriesCapture)
+
+	if int(reqLimit) > maxEntriesLimit && maxEntriesLimit != 0 {
+		return fmt.Errorf("max entries limit per query exceeded, limit > max_entries_limit (%d > %d)", reqLimit, maxEntriesLimit)
+	}
+	return nil
+}
+
+func validateMatchers(req *http.Request, limits Limits, matchers []*labels.Matcher) error {
+	tenants, err := tenant.TenantIDs(req.Context())
+	if err != nil {
+		return err
+	}
+
+	actual := make(map[string]struct{}, len(matchers))
+	for _, m := range matchers {
+		actual[m.Name] = struct{}{}
+	}
+
+	for _, tenant := range tenants {
+		required := limits.RequiredLabelMatchers(req.Context(), tenant)
+		var missing []string
+		var present []string
+		for _, label := range required {
+			if _, found := actual[label]; !found {
+				missing = append(missing, label)
+			} else {
+				present = append(present, label)
+			}
+		}
+
+		if len(missing) > 0 {
+			if len(present) > 0 {
+				return fmt.Errorf("stream selector included required matchers %s but is missing %s", strings.Join(present, ", "), strings.Join(missing, ", "))
+			}
+			return fmt.Errorf("stream selector is missing required matchers %s", strings.Join(missing, ", "))
+		}
```

Comment:
Hey sorry I think my original comment here wasn't as clear as I could have made it. I think in the case of any required labels being missing it would be most useful to the user if we included _all_ the labels that were present in the query, not just the ones that were required, in the error message. 

For example. if we have the query 
```
some_aggregation({label_a="x", label_b="y"} | some filter here) / some_aggregation({label_a="a", label_b="s", label_c="d"} | some filter here) / 
```
And the required labels were `[label_a, label_b, label_c]`, if the above query were actually a lot longer (# of characters) it could be hard to see that the left side is missing just one of the required labels and tunnel vision on the fact that the right side has all the required labels. If the error message was `"stream selector is missing required matchers [label_], labels present in the query were [label_a, label_b]"` if might be more obvious where the problem actually is.

Reply:
Hm, I have it similar to your example but will adopt it.

Reply:
I think what might have been confusing is if the left side was missing one of the required labels, but it was `some_aggregation({label_a="x", label_b="y", another_label="xyz"} | some filter here) / right_side` I'd expect the error to be `"stream selector is missing required matchers [label_c], labels present in the query were [label_a, label_b, another_label]"`